### PR TITLE
fix: terraform-plugin-sdk zeros *int fields

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/7582#issuecomment-1568756959

Changes:
* add computed fields `min_ok` and `max_ok` to `Validation`. These fields substitute the impaired logic in terraform-plugin-sdk (can't handle *ptr).
* replace `*int` with `int` (revert last change) as pointers don't give any benefits now.

~~For some reason, the Terraform state contains `validation.0.min = 0`. I managed to fix [parameterDataSource](https://github.com/coder/terraform-provider-coder/blob/main/provider/parameter.go#L61) to handle [nils](https://github.com/coder/terraform-provider-coder/pull/122).~~

~~If you look at the state, it doesn't contain absent root properties of `coder_parameter` (like `display_name`), but it contains all for `Validation`. Why :) ?~~

### CODE REVIEW

File: provider/parameter_test.go [hunk 1]
```diff
@@ -19,6 +19,34 @@ func TestParameter(t *testing.T) {
...
+				"default":          "2",
+				"validation.0.max": "9",
+			} {
+				require.Equal(t, expected, state.Primary.Attributes[key])
+			}
+
+			_, foundDisplayName := state.Primary.Attributes["display_name"]
+			require.False(t, foundDisplayName, "display_name = "+state.Primary.Attributes["display_name"])
+			_, foundValidationMin := state.Primary.Attributes["validation.0.min"]
+			require.False(t, foundValidationMin, "validation.0.min = "+state.Primary.Attributes["validation.0.min"])
```
Comment (mtojek) [inline provider/parameter_test.go:47]: @kylecarbs Maybe you have a clue what is the reason?

File: provider/parameter.go [hunk 2]
```diff
@@ -401,10 +417,10 @@ func valueIsType(typ, value string) diag.Diagnostics {
 
 func (v *Validation) Valid(typ, value string) error {
 	if typ != "number" {
-		if v.Min != nil {
+		if !v.MinDisabled && v.Min != 0 {
 			return fmt.Errorf("a min cannot be specified for a %s type", typ)
 		}
-		if v.Max != nil {
+		if !v.MaxDisabled && v.Max != 0 {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply provider/parameter.go]: Fixed.

File: provider/parameter.go [hunk 3]
```diff
@@ -401,10 +417,10 @@ func valueIsType(typ, value string) diag.Diagnostics {
 
 func (v *Validation) Valid(typ, value string) error {
 	if typ != "number" {
-		if v.Min != nil {
+		if !v.MinDisabled && v.Min != 0 {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply provider/parameter.go]: Fixed.

File: provider/parameter_test.go [hunk 4]
```diff
@@ -495,17 +562,19 @@ func TestValueValidatesType(t *testing.T) {
...
-		Type:  "number",
-		Value: "2",
-		Max:   ptrNumber(1),
-		Error: regexp.MustCompile("is more than the maximum 1"),
+		Name:        "NumberAboveMax",
+		Type:        "number",
+		Value:       "2",
+		Max:         1,
+		MinDisabled: true,
+		Error:       regexp.MustCompile("is more than the maximum 1"),
```
[the remark this answers is not in the document]
Comment (mtojek) [reply provider/parameter_test.go:672]: I added more test cases (Min, Max, MinZero, MaxZero, LesserThan, GreaterThan, NotInRange).

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**